File: Java/src/tm2c/DBConnection.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package tm2c;

import java.sql.*;
import java.util.ArrayList;

/**
 *
 * @author csp98
 */
public class DBConnection {

    private static final String DB_URL = Parameters.DB_URL;
    private static final String DB_DRV = Parameters.DB_DRV;
    private static final String DB_USER = Parameters.DB_USER;
    private static final String DB_PASSWORD = Parameters.DB_PASSWORD;
    private Connection connection = null;
    private Statement statement = null;
    private String username = null;

    public DBConnection() {
        try {
            Class.forName(DB_DRV);
            connection = DriverManager.getConnection(
                    DB_URL, DB_USER, DB_PASSWORD);
            statement = connection.createStatement();
        } catch (ClassNotFoundException | SQLException ex) {
            System.out.println(ex);
        }
    }

    public boolean login(String username, String password) {
        boolean success = isValidUser(username, password);
        if (success) {
            this.username = username;
        }
        return success;
    }

    public void logout() {
        username = null;
    }

    private boolean isValidUser(String username, String password) {
        boolean valid = false;
        ResultSet rs;
        try {
            rs = statement.executeQuery("select * from student where username=\""
                    + username + "\" and password=sha1(md5(\"" + password + "\"));");
            if (rs.next()) {
                valid = true;
            }
        } catch (SQLException e) {
            System.out.println(e);
        }

        return valid;
    }

    public String getLocation(String building) {
        ResultSet query;
        String result = "";
        try {
            query = statement.executeQuery("SELECT location from building where building_code=\"" + building + "\";");
            if (query.next()) {
                result = query.getString(1);
            }
        } catch (SQLException e) {
            System.out.println(e);
        }
        return result;
    }

    public boolean createUser(String username, String password, String name, String email) {
        boolean success = false;
        try {
            statement.execute("INSERT INTO student(username,password,name,email) VALUES (\""
                    + username + "\",sha1(md5(\"" + password + "\")),\"" + name + "\",\"" + email + "\");");
            success = true;
        } catch (SQLException e) {
            if (!e.toString().contains("Duplicate")) {
                System.out.println(e);
            }
        }
        this.username = username;
        return success;
    }

    public boolean addCourse(Course c) {
        String start_time = c.getStartTime();
        String end_time = c.getEndTime();
        String course_name = c.getCourseName();
        String course_code = c.getCourseCode();
        String room_code = c.getRoomCode();
        String building_code = c.getBuildingCode();
        String building_location = c.getBuildingLocation();
        String day = c.getDay();
        day = day.toUpperCase();
        boolean success = false;
        if (username != null) {
            try {

                statement.execute("INSERT INTO course"
                        + " SELECT * FROM (SELECT \"" + course_code + "\",\"" + course_name + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT course_code FROM course where course_code=\"" + course_code + "\") LIMIT 1;");
                statement.execute("INSERT INTO room"
                        + " SELECT * FROM (SELECT \"" + room_code + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT room_code FROM room where room_code=\"" + room_code + "\") LIMIT 1;");

                statement.execute("INSERT INTO is_registered"
                        + " SELECT * FROM (SELECT \"" + username + "\",\"" + course_code + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT course_code FROM is_registered where course_code=\"" + course_code + "\") LIMIT 1;");

                statement.execute("INSERT INTO imparted SELECT * FROM(SELECT \"" + course_code
                        + "\",\"" + room_code + "\",\"" + day + "\",\"" + start_time + "\",\"" + end_time
                        + "\") AS tmp WHERE NOT EXISTS (SELECT course_code FROM imparted WHERE course_code=\""
                        + course_code + "\") LIMIT 1;");

                statement.execute("INSERT INTO located" + " SELECT * FROM (SELECT \"" + room_code + "\",\"" + room_code.charAt(0) + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT room_code FROM located where room_code=\"" + course_code + "\") LIMIT 1;");

                success = true;
            } catch (SQLException ex) {
                if (!ex.toString().contains("Duplicate")) {
                    System.out.println(ex);
                }
            }
        } else {
            System.err.println("You must log in first.");
        }
        return success;
    }

    public boolean addCourse(String course_code, String course_name, String room,
            String day, String start_hour, String end_hour) {
        day = day.toUpperCase();
        boolean success = false;
        if (username != null) {
            try {

                statement.execute("INSERT INTO course"
                        + " SELECT * FROM (SELECT \"" + course_code + "\",\"" + course_name + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT course_code FROM course where course_code=\"" + course_code + "\") LIMIT 1;");
                statement.execute("INSERT INTO room"
                        + " SELECT * FROM (SELECT \"" + room + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT room_code FROM room where room_code=\"" + room + "\") LIMIT 1;");

                statement.execute("INSERT INTO is_registered"
                        + " SELECT * FROM (SELECT \"" + username + "\",\"" + course_code + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT course_code FROM is_registered where course_code=\"" + course_code + "\") LIMIT 1;");

                statement.execute("INSERT INTO imparted SELECT * FROM(SELECT \"" + course_code
                        + "\",\"" + room + "\",\"" + day + "\",\"" + start_hour + "\",\"" + end_hour
                        + "\") AS tmp WHERE NOT EXISTS (SELECT course_code FROM imparted WHERE course_code=\""
                        + course_code + "\") LIMIT 1;");

                statement.execute("INSERT INTO located" + " SELECT * FROM (SELECT \"" + room + "\",\"" + room.charAt(0) + "\") AS tmp WHERE NOT EXISTS"
                        + "(SELECT room_code FROM located where room_code=\"" + course_code + "\") LIMIT 1;");

                success = true;
            } catch (SQLException ex) {
                if (!ex.toString().contains("Duplicate")) {
                    System.out.println(ex);
                }
            }
        } else {
            System.err.println("You must log in first.");
        }
        return success;
    }

    public ArrayList<Course> getAllCourses() {
        ResultSet query = null;
        ArrayList<Course> result = new ArrayList();
        Course aux;
        if (username != null) {
            try {
                query = statement.executeQuery("SELECT imparted.course_code,building.location,"
                        + "imparted.start_time,imparted.end_time,course.name,imparted.day,building.building_code,imparted.room_code FROM is_registered,"
                        + "building,located,student,course,imparted WHERE student.username=\"" + username + "\"\n"
                        + "AND course.course_code=imparted.course_code AND student.username=is_registered.username AND is_registered.course_code=imparted.course_code \n"
                        + "AND located.room_code=imparted.room_code AND building.building_code=located.building_code;");
                System.out.println("SELECT imparted.course_code,building.location,"
                        + "imparted.start_time,imparted.end_time,course.name,imparted.day,building.building_code,imparted.room_code FROM is_registered,"
                        + "building,located,student,course,imparted WHERE student.username=\"" + username + "\"\n"
                        + "AND course.course_code=imparted.course_code AND student.username=is_registered.username AND is_registered.course_code=imparted.course_code \n"
                        + "AND located.room_code=imparted.room_code AND building.building_code=located.building_code;");
                String start_time;
                String end_time;
                String course_name;
                String course_code;
                String room_code;
                String building_code;
                String building_location;
                String day;
                while (query.next()) {
                    start_time = query.getString(3);
                    end_time = query.getString(4);
                    course_name = query.getString(5);
                    course_code = query.getString(1);
                    room_code = query.getString(8);
                    building_code = query.getString(7);
                    building_location = query.getString(2);
                    day = query.getString(6);

                    aux = new Course(day, start_time, end_time, course_name, course_code, room_code, building_location);
                    result.add(aux);
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        } else {
            System.err.println("You must log in first.");
        }
        return result;
    }

    public boolean removeCourse(String course_code) {
        boolean success = false;
        if (username != null) {
            try {
                statement.execute("DELETE FROM is_registered WHERE course_code=\"" + course_code + "\" AND username=\"" + username + "\";");
                success = true;
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        } else {
            System.err.println("You must log in first.");
        }
        return success;
    }

    public Course getCourseAt(String time, String day) {
        day = day.toUpperCase();
        ResultSet query = null;
        Course result = null;
        String start_time;
        String end_time;
        String course_name;
        String course_code;
        String room_code;
        String building_code;
        String building_location;
        if (username != null) {
            try {
                query = statement.executeQuery("SELECT distinct imparted.course_code,building.location,"
                        + "imparted.start_time,imparted.end_time,course.name,imparted.day,building.building_code,imparted.room_code"
                        + " FROM course,imparted,located,building,is_registered WHERE "
                        + "located.building_code=building.building_code AND"
                        + " is_registered.username=\"" + username + "\" AND located.room_code=imparted.room_code AND "
                        + "imparted.course_code=course.course_code AND imparted.course_code=is_registered.course_code AND imparted.start_time<=\"" + time + "\" "
                        + "AND imparted.end_time>\"" + time + "\" "
                        + "AND imparted.day=\"" + day + "\" ORDER BY start_time;"
                );
                System.out.println("SELECT distinct imparted.course_code,building.location,"
                        + "imparted.start_time,imparted.end_time,course.name,imparted.day,building.building_code,imparted.room_code"
                        + " FROM course,imparted,located,building,is_registered WHERE "
                        + "located.building_code=building.building_code AND"
                        + " is_registered.username=\"" + username + "\" AND located.room_code=imparted.room_code AND "
                        + "imparted.course_code=course.course_code AND imparted.course_code=is_registered.course_code AND imparted.start_time<=\"" + time + "\" "
                        + "AND imparted.end_time>\"" + time + "\" "
                        + "AND imparted.day=\"" + day + "\";");
                while (query.next()) {
                    start_time = query.getString(3);
                    end_time = query.getString(4);
                    course_name = query.getString(5);
                    course_code = query.getString(1);
                    room_code = query.getString(8);
                    building_code = query.getString(7);
                    building_location = query.getString(2);
                    result = new Course(day, start_time, end_time, course_name, course_code, room_code, building_location);
                }
            } catch (SQLException ex) {

                System.out.println(ex);
            }
        } else {
            System.err.println("You must log in first.");
        }
        return result;
    }

    public Course getNextCourse(String time, String day) {
        day = day.toUpperCase();
        ResultSet query = null;
        Course result = null;
        String start_time;
        String end_time;
        String course_name;
        String course_code;
        String room_code;
        String building_code;
        String building_location;
        if (username != null) {
            try {
                query = statement.executeQuery("SELECT distinct imparted.course_code,building.location,"
                        + "imparted.start_time,imparted.end_time,course.name,imparted.day,building.building_code,imparted.room_code"
                        + " FROM course,imparted,located,building,is_registered WHERE "
                        + "course.course_code=is_registered.course_code AND located.building_code=building.building_code AND"
                        + " is_registered.username=\"" + username + "\" AND located.room_code=imparted.room_code AND "
                        + "imparted.course_code=is_registered.course_code AND imparted.start_time>\"" + time + "\" "
                        + "AND imparted.day=\"" + day + "\" ORDER BY start_time;"
                );

                while (query.next()) {
                    start_time = query.getString(3);
                    end_time = query.getString(4);
                    course_name = query.getString(5);
                    course_code = query.getString(1);
                    room_code = query.getString(8);
                    building_code = query.getString(7);
                    building_location = query.getString(2);
                    result = new Course(day, start_time, end_time, course_name, course_code, room_code, building_location);
                }
            } catch (SQLException ex) {
                System.out.println(ex);
            }
        } else {
            System.err.println("You must log in first.");
        }
        return result;
    }

    public boolean removeAllCourses() {
        boolean success = false;
        if (username != null) {
            try {
                statement.execute("DELETE FROM is_registered WHERE username=\"" + username + "\";");
                success = true;
            } catch (SQLException ex) {
            }
        } else {
            System.err.println("You must log in first.");
        }
        return success;
    }
}

File: Python/db_connection.py
import mysql.connector
import parameters
import sys
import course


class DBConnection:
    def __init__(self):
        self.db = mysql.connector.connect(
            host=parameters.DB_URL,
            user=parameters.DB_USER,
            passwd=parameters.DB_PASSWORD,
            database=parameters.DB_NAME
        )
        self.username = None

    def is_valid_user(self, username):
        cursor = self.db.cursor()
        valid = False
        cursor.execute(
            "SELECT * FROM student WHERE username=\"{}\";".format(username))
        cursor.fetchall()
        if cursor.rowcount == 1:
            valid = True
        return valid

    def login(self, username):
        success = self.is_valid_user(username)
        if success:
            self.username = username
        return success

    def logout(self):
        self.username = None

    def get_location(self, building):
        cursor = self.db.cursor()
        cursor.execute(
            "SELECT location FROM building WHERE building_code=\"{}\";".format(building))
        result = cursor.fetchall()[0][0]
        return result

    def create_user(self, username, name):
        success = False
        try:
            cursor = self.db.cursor()
            cursor.execute("INSERT INTO student(username,name) VALUES (\"{}\",\"{}\") ".format(
                username, name))
            self.db.commit()
            success = True
        except Exception:
            print("The user already exists.")
        return success

    def add_course(self, c):
        success = False
        start_time = c.start_time
        end_time = c.end_time
        course_name = c.course_name
        room_code = c.room_code
        building_code = c.building_code
        building_location = c.building_location
        day = c.day

        if self.username != None:
            cursor = self.db.cursor()
            cursor.execute(
                "INSERT INTO course SELECT * FROM (SELECT \"{0}\") AS tmp WHERE NOT EXISTS (SELECT course_name FROM course WHERE course_name=\"{0}\") LIMIT 1;".format(course_name))
            cursor.execute(
                "INSERT INTO room SELECT * FROM (SELECT \"{0}\") AS tmp WHERE NOT EXISTS (SELECT room_code FROM room WHERE room_code=\"{0}\") LIMIT 1;".format(room_code))
            cursor.execute(
                "INSERT INTO is_registered SELECT * FROM (SELECT \"{0}\",\"{1}\") AS tmp WHERE NOT EXISTS (SELECT course_name FROM is_registered WHERE course_name=\"{1}\") LIMIT 1;".format(self.username, course_name))

            cursor.execute("INSERT INTO imparted SELECT * FROM (SELECT \"{0}\",\"{1}\",\"{2}\",\"{3}\",\"{4}\") AS tmp WHERE NOT EXISTS (SELECT course_name FROM imparted WHERE course_name=\"{0}\") LIMIT 1;".format(
                course_name, room_code, day, start_time, end_time))

            cursor.execute(
                "INSERT INTO located SELECT * FROM (SELECT \"{0}\",\"{1}\") AS tmp WHERE NOT EXISTS (SELECT room_code FROM located WHERE room_code=\"{0}\") LIMIT 1;".format(room_code, building_code))

            self.db.commit()
            success = True
        return success

    def get_all_courses(self):
        result = []
        print(self.username)
        if self.username != None:
            cursor = self.db.cursor(dictionary=True)
            cursor.execute("SELECT * FROM is_registered,building,located,student,course,imparted WHERE student.username=\"{}\"AND course.course_name=imparted.course_name AND student.username=is_registered.username AND is_registered.course_name=imparted.course_name AND located.room_code=imparted.room_code AND building.building_code=located.building_code;".format(self.username))
            row = cursor.fetchone()
            while row != None:
                aux = course.Course(course_name=row['course_name'], start_time=row['start_time'],
                                    end_time=row['end_time'], day=row['day'], room_code=row['room_code'])
                result.append(aux)
                row = cursor.fetchone()
        return result

    def remove_course(self, course_name):
        success = False
        if self.username != None:
            cursor = self.db.cursor()
            try:
                cursor.execute("DELETE FROM is_registered WHERE username=\"{}\" AND course_name=\"{}\";".format(
                    self.username, course_name))
                self.db.commit()
                success = True
            except Exception:
                pass
        return success

    def remove_all_courses(self):
        success = False
        if self.username != None:
            cursor = self.db.cursor()
            try:
                cursor.execute(
                    "DELETE FROM is_registered WHERE username=\"{}\"".format(self.username))
                self.db.commit()
                success = True
            except Exception:
                pass
        return success

    def get_next_course(self,time, day):
        day=day.upper()
        result = None
        cursor = self.db.cursor(dictionary=True)
        cursor.execute("SELECT distinct imparted.course_name,building.location,imparted.start_time,imparted.end_time,course.course_name,imparted.day,building.building_code,imparted.room_code FROM course,imparted,located,building,is_registered WHERE course.course_name=is_registered.course_name AND located.building_code=building.building_code AND is_registered.username=\"{0}\" AND located.room_code=imparted.room_code AND imparted.course_name=is_registered.course_name AND imparted.start_time>\"{1}\" AND imparted.day=\"{2}\" ORDER BY start_time;".format(
            self.username, time, day))
        data = cursor.fetchone()
        if data==None:
            result=None
        else:
            result = course.Course(course_name=data['course_name'], start_time=data['start_time'],
                                   end_time=data['end_time'], day=data['day'], room_code=data['room_code'])

        return result
"""
obj = DBConnection()
print(obj.login("Juan", "pass"))
# print(obj.get_location("I"))
#obj.create_user("lolo", "x", "Lolo", "mail")
#print(obj.is_valid_user("lolo", "x"))
co = course.Course("name", "code", "12:00", "13:00",
                   "monday", "IB231", obj.get_location("I"))
obj.add_course(co)
e = obj.get_all_courses()
for x in e:
    print(x.to_s())

obj.remove_course("code")
print("code borrado")
v = obj.get_all_courses()
for x in v:
    print(x.to_s())

obj.get_next_course(day="monday",time="2:00")
"""
